dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/tei/query/OrganisationUnitCondition.java

```diff
@@ -70,20 +92,86 @@ public class OrganisationUnitCondition extends BaseRenderable
+
+        if ( ouMode == SELECTED )
+        {
+
+            List<String> items = organisationUnits.stream()
+                .map( BaseIdentifiableObject::getUid )
+                .collect( Collectors.toList() );
+
+            return isEmpty( items ) ? FALSE_CONDITION
+                : BinaryConditionRenderer.of(
+                    Field.ofDimensionIdentifier( dimensionIdentifier ),
+                    IN,
+                    items,
+                    STRING,
+                    queryContext );
+
+        }
+        else if ( ouMode == CHILDREN )
+        {
+
+            List<String> items = organisationUnits.stream()
+                .map( OrganisationUnit::getChildren )
+                .flatMap( Collection::stream )
+                .map( BaseIdentifiableObject::getUid )
+                .collect( Collectors.toList() );
+
+            return isEmpty( items ) ? FALSE_CONDITION
+                : BinaryConditionRenderer.of(
+                    Field.ofDimensionIdentifier( dimensionIdentifier ),
+                    IN,
+                    items,
+                    STRING,
+                    queryContext );
+        }
+
+        // ouMode = Descendants
         List<Renderable> orgUnitConditions = new ArrayList<>();
 
-        for ( DimensionParamItem item : dimensionIdentifier.getDimension().getItems() )
+        for ( OrganisationUnit organisationUnit : organisationUnits )
         {
-            BinaryConditionRenderer condition = BinaryConditionRenderer.of(
-                Field.ofDimensionIdentifier( dimensionIdentifier ),
+            orgUnitConditions.add( BinaryConditionRenderer.of(
+                Field.ofRenamedDimensionIdentifier(
+                    dimensionIdentifier,
+                    OULEVEL + organisationUnit.getLevel() ),
                 IN,
-                item.getValues(),
+                List.of( organisationUnit.getUid() ),
                 STRING,
-                queryContext );
-            orgUnitConditions.add( condition );
+                queryContext ) );
         }
 
-        return AndCondition.of( orgUnitConditions ).render();
+        return isEmpty( orgUnitConditions ) ? FALSE_CONDITION : OrCondition.of( orgUnitConditions );
     }
 
+    private OrganisationUnitSelectionMode getOuMode()
+    {
+        return Optional.of( queryContext )
+            .map( QueryContext::getTeiQueryParams )
+            .map( TeiQueryParams::getCommonParams )
+            .map( CommonParams::getOuMode )
+            .filter( ACCEPTED_OU_MODES::contains )
+            .orElse( DEFAULT_ORG_UNIT_SELECTION_MODE );
+    }
+
+    private List<OrganisationUnit> getOrganisationUnits()
+    {
+        return Optional.of( dimensionIdentifier )
+            .map( DimensionIdentifier::getDimension )
+            .map( DimensionParam::getDimensionalObject )
+            .map( DimensionalObject::getItems )
+            .orElse( Collections.emptyList() )
```

Comment:
We usually prefer `List.of()` for ready-only situations.

Reply:
I think emptyList as well is immutable

Reply:
> I think emptyList as well is immutable

I'll replace it with `List.of()`